### PR TITLE
make the loss context manager easier to extend

### PR DESCRIPTION
**Pull request by winglian**
# What does this PR do?

Using ExitStack makes it easier to extend the functionality of the context manager wrapping the `model(inputs)` call. This can make it easier to add other things like logging contexts, checkpointing contexts, etc down the road.

e.g.

```python
class CustomTrainer(Trainer):
    def compute_loss_context_manager(self):
        ctx_stack = super(). compute_loss_context_manager()
        my_custom_ctx = CustomContext(...)
        ctx_stack.enter_context(my_custom_ctx)
        return ctx_stack
````

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@SunMarc @ArthurZucker 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3695,7 +3695,13 @@ def compute_loss_context_manager(self):
...
         A helper wrapper to group together context managers.
         """
-        return self.autocast_smart_context_manager()
+        ctx_stack = contextlib.ExitStack()
+
+        autocast_ctx = self.autocast_smart_context_manager()
+        if not isinstance(autocast_ctx, contextlib.nullcontext):
+            ctx_stack.enter_context(autocast_ctx)
+
+        return ctx_stack
```
Comment (SunMarc) [inline src/transformers/trainer.py:3704]: nice ! maybe add in the doctstring that users can easily extend the context manager also !

**Review by SunMarc — APPROVED** [review]
Thanks !

**Review by ArthurZucker — APPROVED** [review]
+1 on documenting and thanks!
